src/PolicyEditor.js

```diff
@@ -496,6 +535,116 @@ export class PolicyEditor extends Component {
                             </Row>
                         </Accordion.Body>
                     </Accordion.Item>
+                    <Accordion.Item eventKey="snapshot-actions">
+                        <Accordion.Header><FontAwesomeIcon icon={faFileAlt} />&nbsp;Snapshot Actions</Accordion.Header>
+                        <Accordion.Body>
+                            <SectionHeaderRow />
+                            <Row>
+                                <LabelColumn name="Before Snapshot" help="Script to run before snapshot." />
+                                <WideValueColumn>{OptionalFieldNoLabel(this, "", "policy.actions.beforeSnapshotRoot.path", {})}</WideValueColumn>
+                                {EffectiveValue(this, "actions.beforeSnapshotRoot.path")}
+                            </Row>
+                            <Row>
+                                <LabelColumn name="Timeout" help="Timeout in seconds before Kopia kills the process." />
+                                <WideValueColumn>{OptionalNumberField(this, "", "policy.actions.beforeSnapshotRoot.timeout", {})}</WideValueColumn>
+                                {EffectiveValue(this, "actions.beforeSnapshotRoot.timeout")}
+                            </Row>
+                            <Row>
+                                <LabelColumn name="Command Mode" help="essential (must succeed, default behavior), optional (failures are tolerated) or async (kopia will start the action but not wait for it to finish)." />
+                                <WideValueColumn>
+                                    <Form.Control as="select" size="sm"
+                                        name="policy.actions.beforeSnapshotRoot.mode"
+                                        onChange={this.handleChange}
+                                        value={stateProperty(this, "policy.actions.beforeSnapshotRoot.mode")}>
+                                        <option value="essential">essential</option>
+                                        <option value="optional">optional</option>
+                                        <option value="async">async</option>
+                                    </Form.Control>
+                                </WideValueColumn>
+                                {EffectiveValue(this, "actions.beforeSnapshotRoot.mode")}
+                            </Row>
+                            <hr />
+                            <Row>
+                                <LabelColumn name="After Snapshot" help="Script to run after snapshot." />
+                                <WideValueColumn>{OptionalFieldNoLabel(this, "", "policy.actions.afterSnapshotRoot.path", {})}</WideValueColumn>
+                                {EffectiveValue(this, "actions.afterSnapshotRoot.path")}
+                            </Row>
+                            <Row>
```

Comment:
With actions it's also possible to embed a script `policy.actions.afterSnapshotRoot.script` - can/should we add a textarea here? 

This makes the UI slightly confusing because you can't specify both `script` and `path`. Here's a slightly more complex UI idea, which might be more self-explanatory:

For each section (Before Folder, After Folder, Before Snapshot, After Snapshot) we present a dropdown:

Action Before Folder:

* none
* run external command
* run script

When the selection is `none`, we hide all other controls.
When it is `external command`, we show `path`, `args`, `timeout` and `mode`.
When it is `run script` we show `script`, `timeout` and `mode`

If we implement this dynamic hiding, perhaps we can combine all 4 actions into one Accordion tab?

We should perhaps also add a link to https://kopia.io/docs/advanced/actions/

Reply:
Let's clean up the redundant template code above before addressing this.

Reply:
I think for this to work, I would ideally need to store the action type as part of the policy, but I don't think I can do that, can I? If not, I'm not sure if I can derive the type from set/unset script command or path...